### PR TITLE
impr: moved Mutable parsing and handling as part of the Shadow annotation

### PR DESCRIPTION
**Pull request by 0xJoeMama**

### CODE REVIEW

File: micromixin-transformer/src/main/java/org/stianloader/micromixin/transform/internal/annotation/MixinShadowAnnotation.java [hunk 1]
```diff
@@ -107,14 +121,20 @@ private void apply(@NotNull MixinFieldStub source, @NotNull ClassNode target, @N
...
     public void apply(@NotNull ClassNode to, @NotNull HandlerContextHelper hctx,
-            @NotNull MixinStub sourceStub, @NotNull T source, @NotNull SimpleRemapper remapper,
-            @NotNull StringBuilder sharedBuilder) {
-        // NOP
+                      @NotNull MixinStub sourceStub, @NotNull T source, @NotNull SimpleRemapper remapper,
+                      @NotNull StringBuilder sharedBuilder) {
+        if (!this.isMutable) return;
+
+        if (source instanceof MixinFieldStub) {
+            ((MixinFieldStub) source).field.access &= ~Opcodes.ACC_FINAL;
```
Comment (Geolykt) [inline micromixin-transformer/src/main/java/org/stianloader/micromixin/transform/internal/annotation/MixinShadowAnnotation.java:129]: `source` in this case refers to the original mixin class. This is in contrast to the target (here it is `to`) which is the node we actually want to transform

File: micromixin-annotations/src/main/java/org/spongepowered/asm/mixin/Mutable.java [hunk 2]
```diff
@@ -1,9 +1,17 @@
...
 
 import java.lang.annotation.ElementType;
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
 import java.lang.annotation.Target;
 
-// TODO: Implement for methods
-@Target({ElementType.FIELD})
+/**
+ * {@link Mutable} can only be applied to {@link Shadow}ed fields or methods.
```
Comment (Geolykt) [inline micromixin-annotations/src/main/java/org/spongepowered/asm/mixin/Mutable.java:9]: A minor nitpick, but the linking conventions around here would mean that you use `{@link Shadow shadowed}`. Although this is more of a personal choice I suppose

File: micromixin-transformer/src/main/java/org/stianloader/micromixin/transform/internal/annotation/MixinShadowAnnotation.java [hunk 3]
```diff
@@ -18,23 +20,24 @@
...
     }
 
     @NotNull
-    public static <T0 extends ClassMemberStub> MixinShadowAnnotation<T0> parse(@NotNull AnnotationNode annotation) {
+    public static <T0 extends ClassMemberStub> MixinShadowAnnotation<T0> parse(@NotNull AnnotationNode annotation, @Nullable List<AnnotationNode> allAnnotations) {
         String prefix = "shadow$";
         List<String> aliases = Collections.emptyList();
         if (annotation.values == null) {
-            return new MixinShadowAnnotation<T0>(prefix, aliases);
+            return new MixinShadowAnnotation<T0>(prefix, aliases, false);
```
Comment (Geolykt) [inline micromixin-transformer/src/main/java/org/stianloader/micromixin/transform/internal/annotation/MixinShadowAnnotation.java:40]: If the `@Shadow` has no prefix or alias, then `@Mutable` will get discarded. This is probably not what you want to achieve.

File: micromixin-transformer/src/main/java/org/stianloader/micromixin/transform/internal/annotation/MixinShadowAnnotation.java [hunk 4]
```diff
@@ -52,7 +55,18 @@ public static <T0 extends ClassMemberStub> MixinShadowAnnotation<T0> parse(@NotN
                 throw new MixinParseException("Unimplemented option for @Shadow: " + name);
             }
         }
-        return new MixinShadowAnnotation<T0>(prefix, aliases);
+
+        boolean isMutable = false;
+        if (allAnnotations != null) {
+            for (AnnotationNode currAnn : allAnnotations) {
+                if (currAnn.desc.equals("Lorg/spongepowered/asm/mixin/Mutable;")) {
```
Comment (Geolykt) [inline micromixin-transformer/src/main/java/org/stianloader/micromixin/transform/internal/annotation/MixinShadowAnnotation.java:62]: Would be great if you were to throw a `MixinParseException` if `!currAnn.values.isEmpty()`

**Review by Geolykt — CHANGES_REQUESTED** [review]

File: micromixin-transformer/src/main/java/org/stianloader/micromixin/transform/internal/annotation/MixinShadowAnnotation.java [hunk 5]
```diff
@@ -107,14 +124,24 @@ private void apply(@NotNull MixinFieldStub source, @NotNull ClassNode target, @N
...
-            @NotNull MixinStub sourceStub, @NotNull T source, @NotNull SimpleRemapper remapper,
-            @NotNull StringBuilder sharedBuilder) {
-        // NOP
+                      @NotNull MixinStub sourceStub, @NotNull T source, @NotNull SimpleRemapper remapper,
+                      @NotNull StringBuilder sharedBuilder) {
+        if (!this.isMutable) return;
+
+        if (source instanceof MixinFieldStub) {
+            for (FieldNode fn : to.fields) {
+                if (fn.name.equals(source.getName()) && fn.desc.equals(source.getDesc())) {
```
Comment (Geolykt) [inline micromixin-transformer/src/main/java/org/stianloader/micromixin/transform/internal/annotation/MixinShadowAnnotation.java:133]: This does not properly account for prefixes and aliases. Use the `remapper` instead on the source.getName/source.getDesc

**Review by Geolykt — CHANGES_REQUESTED** [review]

File: micromixin-transformer/src/main/java/org/stianloader/micromixin/transform/internal/annotation/MixinShadowAnnotation.java (hunk 5 above)
Comment (0xJoeMama) [reply micromixin-transformer/src/main/java/org/stianloader/micromixin/transform/internal/annotation/MixinShadowAnnotation.java:133]: Am I supposed to access the fieldRenames field and use the methods from there? If yes, does that method have to take in the remapped or the original field descriptor?

Comment (Geolykt) [reply micromixin-transformer/src/main/java/org/stianloader/micromixin/transform/internal/annotation/MixinShadowAnnotation.java:133]: oh, and it's always the source descriptor

**Review by Geolykt — APPROVED** [review]
okay that looks all correct at this point in time. Do you have anything else to add?

**Comment by 0xJoeMama** [comment]
For now, no not really. But do expect some Accessor and Invoker PRs soon-ish(if not within the next two weeks, sadly it'll be June when I get started on them, no promises made).

**Pull request merged**